lib/ssl_compat/ssl_compat.h

```diff
@@ -0,0 +1,38 @@
+#ifndef _SSL_COMPAT_H_
+#define _SSL_COMPAT_H_
+
+#include <stddef.h>
+
+#include <openssl/ssl.h>
+
+/**
+ * ssl_compat_write_ex(ssl, buf, num, written):
+ * Write ${num} bytes from ${buf} to the ${ssl} connection.  Store the number
+ * of bytes written in ${written}.
+ *
+ * On success, return 1.  On failure, return 0 or a negative value (for some
+ * versions of OpenSSL, the exact value must be passed to SSL_get_err() in
+ * order to receive a meaningful error code).
+ *
+ * COMPATIBILITY: provides SSL_write_ex().  Exception: if you specify a value
```

Comment:
You don't need to mention the `INT_MAX` bit, since `SSL_write_ex` and `SSL_read_ex` are already allowed to perform short writes/reads.  (Or rather, `SSL_read_ex` always is, and `SSL_write_ex` is as long as `SSL_MODE_ENABLE_PARTIAL_WRITE` is turned on, which we do.)

Reply:
How does this work out when the BIO is **blocking**?  As far as I can tell, we have

> If the underlying BIO is blocking, a read function will only return once the read operation has been finished or an error occurred...

> SSL_read_ex() and SSL_peek_ex() will return 1 for success or 0 for failure. Success means that 1 or more application data bytes have been read from the SSL connection.

> https://www.openssl.org/docs/man1.1.1/man3/SSL_read_ex.html

Now, I know that we're using non-blocking in `network_ssl.c`.  But `ssl_compat.c` doesn't know about that.  Similar concern for `SSL_MODE_ENABLE_PARTIAL_WRITE` -- that's set at a different level than `ssl_compatl.c`.

When it comes to ssl, I'd rather be more explicit about our assumptions, not less.  But if you want me to delete everything after "provides SSL_write_ex()", I'll go ahead and do that.

Reply:
Hmm, good point.  I'm actually inclined to move these bits to `lib/network_ssl/network_ssl_compat.c` so that it's (a) clear what they're for and (b) justifiable to make assumptions about how they're going to be used.

Reply:
ok, will do.  But do you think it's worth un-expliciting those assumptions in the function comments for  `ssl_compat_write_ex` ?

I mean if we just write `COMPATIBILITY: does the same thing as SSL_write_ex()`, then we'd be strictly speaking false.  Which is not a good position to be in with SSL stuff!

(I'm currently reading section 3 of "The most dangerous code in the world".)

Reply:
ah, and I now realize that you also mean "... and get rid of the `.h` file".

I still think that explicit assumption in the comment of the `.c` file are better than implied assumptions based on how it's used in `network_ssl.c`.